chat/utils.go

```diff
@@ -33,3 +55,104 @@ func (c *Chat) refreshSignedPreKey(idPubKey ed25519.PublicKey) error {
 	return c.userStorage.PutSignedPreKey(idPubKey, signedPreKey)
 
 }
+
+// generate shared secret id
+func sharedSecretID(sender, receiver ed25519.PublicKey, sharedSecretID []byte) (mh.Multihash, error) {
+	b := bytes.NewBuffer(sender)
+	if _, err := b.Write(receiver); err != nil {
+		return nil, err
+	}
+	if _, err := b.Write(sharedSecretID); err != nil {
+		return nil, err
+	}
+	return mh.Sum(b.Bytes(), mh.SHA3_256, -1)
+}
+
+// create identification
+func sharedSecretInitID(sender, receiver ed25519.PublicKey, msg bpb.ChatMessage) (mh.Multihash, error) {
+	if len(sender) != 32 {
+		return nil, errors.New("sender must be 32 bytes long")
+	}
+	if len(receiver) != 32 {
+		return nil, errors.New("receiver must be 32 bytes long")
+	}
+	b := bytes.Buffer{}
+	if _, err := b.Write(sender); err != nil {
+		return nil, err
+	}
+	if _, err := b.Write(receiver); err != nil {
+		return nil, err
+	}
+	if _, err := b.Write(msg.SenderChatIDKey); err != nil {
+		return nil, err
+	}
+	if _, err := b.Write(msg.SignedPreKey); err != nil {
+		return nil, err
+	}
+	if _, err := b.Write(msg.OneTimePreKey); err != nil {
+		return nil, err
+	}
+	return mh.Sum(b.Bytes(), mh.SHA3_256, -1)
+}
+
+// hash message
+func hashChatMessage(msg bpb.ChatMessage) (mh.Multihash, error) {
+
+	b := bytes.NewBuffer(nil)
+
+	writes := []func(b *bytes.Buffer) (int, error){
+		func(b *bytes.Buffer) (int, error) {
+			return b.Write(msg.OneTimePreKey)
+		},
+		func(b *bytes.Buffer) (int, error) {
+			return b.Write(msg.SignedPreKey)
+		},
+		func(b *bytes.Buffer) (int, error) {
+			return b.Write(msg.EphemeralKey)
+		},
+		func(b *bytes.Buffer) (int, error) {
+			return b.Write(msg.EphemeralKeySignature)
+		},
+		func(b *bytes.Buffer) (int, error) {
+			return b.Write(msg.SenderChatIDKey)
+		},
+		func(b *bytes.Buffer) (int, error) {
+			return b.Write(msg.SenderChatIDKeySignature)
+		},
+		func(b *bytes.Buffer) (int, error) {
+			return b.Write(msg.Message.DoubleRatchetPK)
+		},
+		func(b *bytes.Buffer) (int, error) {
+			n := make([]byte, 4)
+			binary.BigEndian.PutUint32(n, uint32(msg.Message.N))
+			return b.Write(n)
+		},
+		func(b *bytes.Buffer) (int, error) {
+			pn := make([]byte, 4)
+			binary.BigEndian.PutUint32(pn, uint32(msg.Message.Pn))
+			return b.Write(pn)
+		},
+		func(b *bytes.Buffer) (int, error) {
+			return b.Write(msg.Message.CipherText)
+		},
+		func(b *bytes.Buffer) (int, error) {
+			return b.Write(msg.Receiver)
+		},
+		func(b *bytes.Buffer) (int, error) {
+			return b.Write(msg.Sender)
+		},
+		func(b *bytes.Buffer) (int, error) {
+			return b.Write(msg.MessageID)
+		},
+		func(b *bytes.Buffer) (int, error) {
+			return b.Write(msg.UsedSharedSecret)
+		},
+	}
+
+	for _, action := range writes {
+		action(b)
```

Comment:
Usually we should do error handling on a function that returns an error but as per 
```
https://golang.org/pkg/bytes/#Buffer.Write
```
err is always nil, so no change is needed here.